### PR TITLE
cmake: Minor cleanup and refactoring to prepare for mpi library split

### PR DESCRIPTION
**Pull request by bradking**
These changes are needed for the upcoming split of mpi-specific interfaces into separate libraries, but can be made without doing the split.  I'm doing them in a separate PR to get independent testing and to reduce the size of the upcoming main PR for the split.

* Drop unnecessary or redundant public usage requirements.
* Convert `ADIOS2_HAVE_MPI` checks controlling MPI-specific bindings to `ADIOS2_USE_MPI={0,1}`.
* Factor build of HDF5 engine into separate object libraries for the general-purpose and MPI-specific parts.

### CODE REVIEW

File: source/adios2/common/ADIOSConfig.h.in [hunk 1]
```diff
@@ -25,8 +25,16 @@
...
 @ADIOS2_CONFIG_DEFINES@
 
+/* Everything between here and the note above is programatically generated */
+
+#ifdef ADIOS2_HAVE_MPI
+#  define ADIOS2_USE_MPI 1
+#else
+#  define ADIOS2_USE_MPI 0
+#endif
+
```
Comment (chuckatkins) [inline source/adios2/common/ADIOSConfig.h.in]: I think this needs to be more complex to allow the user to appropriately override it.  Something like:
```c++
#ifndef ADIOS2_USE_MPI
  #ifdef ADIOS2_HAVE_MPI
    #define ADIOS2_USE_MPI 1
  #else
    #define ADIOS2_USE_MPI 0
  #endif
#else
  #if ADIOS2_USE_MPI && !(defined ADIOS2_HAVE_MPI)
    #error "ADIOS was not built with MPI support"
  #endif
#endif
```

Comment (bradking) [reply source/adios2/common/ADIOSConfig.h.in]: As of this PR we don't allow the application to pick yet.  That logic will be updated in the main transition PR.

Comment (bradking) [reply source/adios2/common/ADIOSConfig.h.in]: To clarify: without the library split the application cannot pick.  The ability to pick will be added in the main library split PR.

**Comment by JasonRuonanWang** [comment]
@bradking Just a side question. Is the MPI library split going to need a lot of changes to the MPI specific engines such as IMPI and SSC? I am asking because I am planning to do some refactoring work for SSC metadata management which probably needs to touch a lot of code. Then it might be very painful to deal with conflicts. So if you are going to make a lot of changes to the MPI stuff then I will probably just wait for you to finish before I start.

**Comment by bradking** [comment]
@JasonRuonanWang after this PR is merged the library split PR will mostly just re-organize how sources are assigned to libraries.  Very few changes to the sources themselves are needed.  I've made most of those already in earlier PRs.  You can see the changes in the `mpi-library` branch in my repo, which I've pushed out as WIP occasionally and will eventually be the source branch for the PR.

**Comment by bradking** [comment]
See #2058 for the main library split.

**Pull request merged**